File: Programmers/[60061] 기둥과 보 설치/woong-jae/60061.js
function solution(n, build_frame) {
    const colFrames = new Map(), rowFrames = new Map();
    
    const isValid = (x, y, a, b) => {
        const key = `${x},${y}`;
        if(a) { // 보
            if(b === 0 && !rowFrames.has(key)) return true;
            if(colFrames.has(`${x},${y - 1}`) || colFrames.has(`${x + 1},${y - 1}`)) return true;
            if(rowFrames.has(`${x - 1},${y}`) && rowFrames.has(`${x + 1},${y}`)) return true;
        }
        else { // 기둥
            if(b === 0 && !colFrames.has(key)) return true;
            if(y === 0) return true;
            if(colFrames.has(`${x},${y - 1}`)) return true;
            if(rowFrames.has(`${x - 1},${y}`) || rowFrames.has(`${x},${y}`)) return true;
        }
        return false;
    }
    
    const buildFrame = (x, y, a, b) => {
        const key = `${x},${y}`;
        if(!isValid(x, y, a, b)) return;
        if(a) {
            rowFrames.set(key, a);
        }
        else {
            colFrames.set(key, a);
        }
    }
    
    const deleteFrame = (x, y, a, b) => {
        const key = `${x},${y}`;
        if(a) { // 보
            rowFrames.delete(key);
            if(!isValid(x - 1, y, 1, b) 
               || !isValid(x + 1, y, 1, b)
               || !isValid(x + 1, y, 0, b)
               || !isValid(x, y, 0, b)) {
                rowFrames.set(key, a);
            }
        }
        else { // 기둥
            colFrames.delete(key);
            if(!isValid(x, y + 1, 0, b) 
               || !isValid(x - 1, y + 1, 1, b) 
               || !isValid(x, y + 1, 1, b)) {
                colFrames.set(key, a);
            }
        }
    }
    
    build_frame.forEach(([x, y, a, b]) => {
        if(b) buildFrame(x, y, a, b);
        else deleteFrame(x, y, a, b);
    });
    
    const col = Array.from(colFrames)
    .map(([position, a]) => [...position.split(",").map(e => +e), a])
    const row = Array.from(rowFrames)
    .map(([position, a]) => [...position.split(",").map(e => +e), a])
    
    return col.concat(row).sort((a, b) => {
        if(a[0] !== b[0]) return a[0] - b[0];
        if(a[1] !== b[1]) return a[1] - b[1];
        return a[2] - b[2];
    })
}

File: Programmers/[60061] 기둥과 보 설치/Johoseong/60061.py
def check(n, board):
    for x, y, a in board:
        if a == 0: # 기둥
            if y == 0: continue # 바닥 위
            elif [x, y - 1, 0] in board: continue # 다른 기둥 위
            elif [x - 1, y, 1] in board or [x, y, 1] in board: continue # 보 한쪽 위
            else:
                return False
        else: # 보
            if [x, y - 1, 0] in board or [x + 1, y - 1, 0] in board:
                continue
            if [x - 1, y, 1] in board and [x + 1, y, 1] in board:
                continue
            return False
    return True

def solution(n, build_frame):
    board = [] # 0: 기둥, 1: 보

    for x, y, a, b in build_frame:
        if b == 0: # 삭제
            board.remove([x, y, a])
            flag = check(n, board)
            if flag == False:
                board.append([x, y, a])
        else: # 설치
            board.append([x, y, a])
            flag = check(n, board)
            if flag == False:
                board.remove([x, y, a])
    board.sort()
    return board

print(solution(5, [[0,0,0,1],[2,0,0,1],[4,0,0,1],[0,1,1,1],[1,1,1,1],[2,1,1,1],[3,1,1,1],[2,0,0,0],[1,1,1,0],[2,2,0,1]]))